### ISSUE TITLE
Create firmware bypass routine in Boot Menu

### ISSUE THREAD

**Issue by Wierzbowsky** [RBSC/Carnivore2#75]
Create firmware bypass routine in Boot Menu. More info here:

https://www.msx.org/forum/msx-talk/development/routine-to-bypass-a-firmware?page=2

**Comment by Wierzbowsky** [comment]
Ok, there are 2 solutions for this problem.

For the first one, the hook disabling code is loaded into Carnivore2's main RAM, restore all registers to DefConfig state and return control to the system instead of rebooting. From now on selecting DefConfig or by pressing ESC in the Boot Menu, the built-in firmware is automatically bypassed on all machines listed by gdx (Mitsubishi ML-G1, National FS-4000, FS-4500, FS-4600, FS-4700, Toshiba HX-21I, HX-22I, HX-23I, Sony HB-55P, HB-75D, HB-75P, HB-101, HB-101P, HB-201, HB-201P, HB-F1, HB-F1II, HB-F9P/S and similar). I tested this on my physical Hitachi H3 and on all affected machines in OpenMSX emulator.

[url=https://podrezov.com/temp/bypassfw.zip]https://podrezov.com/temp/bypassfw.zip[/url]

The alternative solution - the special ROM file and its RCP configuration for C2MAN utility, should be started from the Boot Menu with the following configuration settings:

![bypassfw_cfg](https://user-images.githubusercontent.com/16538424/180901882-28a00743-3b9d-4cd4-b4c9-d4f57ba929ac.png)

**Comment by rattboi** [comment]
Hi there.

I've tried solution 2 with my Carnivore2 and Panasonic FS-A1. It does indeed work if I do the following:
1. Use c2man to add BYPASSFW.ROM to my flash, along with the RCP file
2. Enter dual-card setup and configure BYPASSFW in the primary slot of the Slave Slot.

However, I have not yet figured out how to get this to play well with Autostart. It seems like I can't have the dual-slot setup saved. I thought perhaps I needed to make a new Config entry in the menu, but there's no way in a config entry to set the slave slot settings (I think?). I have only had my Carnivore2 for a few days, so this could be user error.

I would be willing to try out solution 1, but the FS-A1 isn't on your list. I scanned through the code in BOOTCMFC.ASM and didn't see anything referring to the entries for Panasonic A1-xx bypass that I do see in BYPASSFW.ASM, so I *think* it's not supported, right?

**Comment by rattboi** [comment]
I am blind and see now that there are actually the lines to support Panasonic FS-A1xx bypass.

```
 	ld	a,#23
	ld	(#CBD8),a	; To bypass the FS-A1 firmare
	ld	(#C3CE),a	; To bypass the FS-A1F firmare
	ld	(#C3D2),a	; To bypass the FS-A1mk2 firmare
```

I'll write this to my card and try it

**Comment by rattboi** [comment]
I've flashed the new bootmenu in the above zip and tried both pressing escape and choosing DefConfig and neither bypassed the firmware. Maybe I misunderstood the steps, as you never called out flashing BOOTCMFC.BIN from the zip, but that was my assumption.

**Comment by Wierzbowsky** [comment]
It works in the emulator, but may not work on a real machine. I will need to do more testing. So, this v2.54 is not going public until I fix all issues.

**Comment by rattboi** [comment]
Let me know of any way I can help. I'm down to test.

**Comment by Wierzbowsky** [comment]
I had to drop the firmware bypass functionality in the Boot Menu because it affects Nextor and some games. So, the only way would be using the special firmware bypass ROM.

**Comment by Wierzbowsky** [comment]
Firmware bypass ROM with its RCP file were released on 14.05.2023
